### PR TITLE
[Datastore] When necessary, save the bucket as part of the data source profile and use the default path if the path is not specified

### PR DESCRIPTION
**Pull request by alxtkr77**
[ML-6326](https://iguazio.atlassian.net/browse/ML-6326)

[ML-6326]: https://iguazio.atlassian.net/browse/ML-6326?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/store/datastore.md [hunk 1]
```diff
@@ -179,6 +179,7 @@ The equivalent to this parameter in environment authentication is "AZURE_STORAGE
 Credential authentication:
 - `credential` &mdash; TokenCredential or SAS token. The credentials with which to authenticate.
 This variable is sensitive information and is kept confidential.
+- `bucket` &mdash; A string representing the bucket. When specified, it is automatically appended to the object path, and thus, it should not be manually included in the target path by the user.
```
Comment (gtopper) [inline docs/store/datastore.md]:
```suggestion
- `bucket` &mdash; A string representing the bucket. When specified, it is automatically prepended to the object path, and thus, it should not be manually included in the target path by the user. 
```
(all instances)

File: mlrun/datastore/datastore_profile.py [hunk 2]
```diff
@@ -185,6 +185,17 @@ class DatastoreProfileS3(DatastoreProfile):
     assume_role_arn: typing.Optional[str] = None
     access_key_id: typing.Optional[str] = None
     secret_key: typing.Optional[str] = None
+    bucket: typing.Optional[str] = None
+
+    @pydantic.validator("bucket")
+    def check_bucket(cls, v):
+        if v is None:
```
Comment (gtopper) [inline mlrun/datastore/datastore_profile.py]:
```suggestion
        if not v:
```
I'm assuming an empty string won't cut it either.
(all instances)

File: mlrun/datastore/datastore_profile.py [hunk 3]
```diff
@@ -283,7 +308,10 @@ def url(self, subpath) -> str:
         if subpath.startswith("/"):
             #  in gcs the path after schema is starts with bucket, wherefore it should not start with "/".
             subpath = subpath[1:]
-        return f"gcs://{subpath}"
+        if self.bucket:
+            return f"gcs://{self.bucket}/{subpath}"
+        else:
+            return f"gcs://{subpath}"
```
Comment (gtopper) [inline mlrun/datastore/datastore_profile.py:319]: Looks like this is an existing issue, but there's something wrong with the subpath being assumed to not start with a slash here, but the opposite assumption is made in S3.

File: mlrun/datastore/targets.py [hunk 4]
```diff
@@ -656,6 +656,41 @@ def get_target_templated_path(self):
...
+                    _get_target_path(
+                        self,
+                        self._resource,
+                        self.run_id is not None,
+                        netloc=parsed_url.netloc,
+                        scheme=parsed_url.scheme,
+                    ),
+                    self.run_id,
+                    is_single_file,
+                )
```
Comment (gtopper) [inline mlrun/datastore/targets.py]: Looks like a coding accident.
```suggestion
```

File: tests/system/datastore/test_aws_s3.py [hunk 5]
```diff
@@ -140,3 +152,40 @@ def test_ingest_with_parquet_source(self, use_datastore_profile):
...
+        df = pd.DataFrame(data)
+        source_path = param["parquet_url"]
+        with tempfile.NamedTemporaryFile(mode="w+", delete=True) as temp_file:
+            df.to_parquet(temp_file.name)
+            path_only = self.s3["s3"]["parquet_url"]
+            s3_fs.put_file(temp_file.name, path_only)
+
+        parquet_source = ParquetSource(name="test", path=source_path)
+
+        # ingest
```
Comment (gtopper) [inline tests/system/datastore/test_aws_s3.py]: Ingest is several lines down, and doesn't really need to be announced.
```suggestion
```

File: tests/system/datastore/test_aws_s3.py [hunk 6]
```diff
@@ -140,3 +152,40 @@ def test_ingest_with_parquet_source(self, use_datastore_profile):
...
+        expected_default_ds_data_prefix = get_default_prefix_for_target(
+            "dsnosql"
+        ).format(
+            ds_profile_name="s3ds_profile_with_bucket",
+            project=fset.metadata.project,
+            kind=targets[0].kind,
+            name=fset.metadata.name,
+        )
+
+        assert fset.get_target_path().startswith(expected_default_ds_data_prefix)
```
Comment (gtopper) [inline tests/system/datastore/test_aws_s3.py:189]: Shouldn't we check that it actually wrote the data also?

File: docs/store/datastore.md (hunk 1 above)
Comment (gtopper) [reply docs/store/datastore.md]: Also, need to add the fact that the user needs to specify this, and will be required to do so from 1.9.0 onward.

File: tests/system/datastore/test_aws_s3.py [hunk 7]
```diff
@@ -140,3 +152,40 @@ def test_ingest_with_parquet_source(self, use_datastore_profile):
         assert_frame_equal(
             df.sort_index(axis=1), result.sort_index(axis=1), check_like=True
         )
+
+    def test_ingest_ds_default_target(self):
+        #  create source
```
Comment (gtopper) [inline tests/system/datastore/test_aws_s3.py]: The source is created 14 lines below, and doesn't really need to be announced.
```suggestion
```

File: mlrun/datastore/datastore_profile.py (hunk 3 above)
Comment (alxtkr77) [reply mlrun/datastore/datastore_profile.py:319]: Like you said - its a separate issue.

Comment (gtopper) [reply mlrun/datastore/datastore_profile.py:319]: I guess at least add a TODO?

File: mlrun/datastore/datastore_profile.py [hunk 8]
```diff
@@ -311,12 +339,26 @@ class DatastoreProfileAzureBlob(DatastoreProfile):
...
     sas_token: typing.Optional[str] = None
     credential: typing.Optional[str] = None
+    bucket: typing.Optional[str] = None
+
+    @pydantic.validator("bucket")
+    def check_bucket(cls, v):
+        if not v:
+            warnings.warn(
+                "The 'bucket' attribute will be mandatory starting from version 1.9",
+                DeprecationWarning,
```
Comment (gtopper) [inline mlrun/datastore/datastore_profile.py]: It turns out that we use `FutureWarning` everywhere else in the code base, so we should probably change this (all instances in this file).
```suggestion
                FutureWarning,
```

File: mlrun/datastore/datastore_profile.py (hunk 3 above)
Comment (gtopper) [reply mlrun/datastore/datastore_profile.py:319]: I guess in S3 though.

**Review by gtopper — APPROVED** [review]

**Pull request merged**